packages/controllers/src/permissions/Permission.ts

```diff
@@ -0,0 +1,463 @@
+import { Json } from 'json-rpc-engine';
+import { nanoid } from 'nanoid';
+import { NonEmptyArray } from '../utils';
+import { CaveatConstraint } from './Caveat';
+// This is used in a docstring, but ESLint doesn't notice it.
+/* eslint-disable @typescript-eslint/no-unused-vars */
+import type { PermissionController } from './PermissionController';
+import type { Caveat } from './Caveat';
+/* eslint-enable @typescript-eslint/no-unused-vars */
+
+/**
+ * The origin of a subject.
+ * Effectively the GUID of an entity that can have permissions.
+ */
+export type OriginString = string;
+
+/**
+ * The name of a permission target.
+ */
+export type GenericTargetName = string;
+
+/**
+ * A `ZCAP-LD`-like permission object. A permission is associated with a
+ * particular `invoker`, which is the holder of the permission. Possessing the
+ * permission grants access to a particular restricted resource, identified by
+ * the `parentCapability`. The use of the restricted resource may be further
+ * restricted by any `caveats` associated with the permission.
+ *
+ * See the documentation for details.
+ */
+export type PermissionConstraint = {
+  /**
+   * The context(s) in which this capability is meaningful.
+   *
+   * It is required by the standard, but we make it optional since there is only
+   * one context in our usage (i.e. the user's MetaMask instance).
+   */
+  readonly '@context'?: NonEmptyArray<string>;
+
+  // TODO:TS4.4 Make optional
+  /**
+   * The caveats of the permission.
+   * @see {@link Caveat} For more information.
+   */
+  readonly caveats: null | NonEmptyArray<CaveatConstraint>;
+
+  /**
+   * The creation date of the permission, in UNIX epoch time.
+   */
+  readonly date: number;
+
+  /**
+   * The GUID of the permission object.
+   */
+  readonly id: string;
+
+  /**
+   * The origin string of the subject that has the permission.
+   */
+  readonly invoker: OriginString;
+
+  /**
+   * A pointer to the resource that possession of the capability grants
+   * access to.
+   *
+   * In the context of MetaMask, this is always the name of an RPC method.
+   */
+  readonly parentCapability: string;
+};
+
+/**
+ * A `ZCAP-LD`-like permission object. A permission is associated with a
+ * particular `invoker`, which is the holder of the permission. Possessing the
+ * permission grants access to a particular restricted resource, identified by
+ * the `parentCapability`. The use of the restricted resource may be further
+ * restricted by any `caveats` associated with the permission.
+ *
+ * See the documentation for details.
+ *
+ * @template TargetKey - They key of the permission target that the permission
+ * corresponds to.
+ * @template AllowedCaveat - A union of the allowed {@link Caveat} types
+ * for the permission.
+ */
+export type Permission<
```

Comment:
Maybe `ValidPermission` would be a better name? Then we wouldn't need to alias it in `PermissionController`. It does seem like it's taking a permission shape and validating two properties.

The fact that this was aliased confused the hell out of me because some TypeScript errors and intellisense tooltips used the non-aliased name.

Reply:
4d24c94